### DISCUSSION TITLE
Moving Forward, and Priorities

### DISCUSSION THREAD

**Discussion by jsamoocha** [hozn/stravalib#340] [Team Posts]
Hi all,

We made great progress in the last period, and I thought it would be a good moment to reflect on what's needed most in stravalib in the (near) future. I'll kick off with my views (which, of course, are just an opinion), and I am curious about your thoughts.

#### Efficient and reliable handling of PRs and API changes
The following tools/processes should facilitate this:

- CI (90% done, code formatting and linting would make this 100% complete)
- Test coverage (I'd like to see 100% coverage for the `client` and `protocol` modules)
- Auto-generated (pydantic) model (10% done, needs a lot of experiments and testing)

#### Documentation
This should show the value of stravalib (over, e.g., auto-generated Swagger clients), with clear examples of:

- Paginated results and their benefits in terms of, e.g., request latency
- Unit conversion
- Implicit “chained” requests (using the bound client)
- Rate limiting

#### Future additions
This library should remain simple, boring, and easy to use. Given that, there may still be useful features to add that I feel some people (cough, cough) have built homegrown solutions for, such as:

- Automatic request retrying (e.g., in case of temporary 5XX responses)
- Automatic token refreshing
- async requests? 🙈

BTW, starting next year, I'll be more constrained time-wise and will likely be less responsive than in the past weeks.

Cheers,

Jonatan

**Comment by lwasser** [comment]
Hey there 👋 
I have both timeline questions and thoughts. I agree with everything above but would add a few things:

* we have numerous issues to work through. ive been chipping away but i think it would be nice to slowly close existing issues. Some may be super old and thus we can close them now?

I'd add to: **Efficient and reliable handling of PRs and API changes**
*  **Code format / linting workflow:** While I know that this item doesn't preclude code from running, it does maintain consistent style. it will also make it easier for contributors to ensure their code meets our style criteria. and that code remains readable. 

## Documentation 
### Work on now? Work on tutorials now maybe? 
1. **API documentation:** Right now we have a good start at API documentation but it's not great. Between the complexity of the API, lack of descriptive / self describing (but short) variable names our API documentation is not complete. Further we're mostly using harder-to-read sphinx rst docstring format vs numpy-style. 
2. "Tutorial" / vignette documentation: I think we need to create some real end to end workflow examples in our tutorials! And i'd love to work on this. If we create a small dataset that contains a client object that might help!   I'd also like to look into implementing [`sphinx-gallery`](https://sphinx-gallery.github.io/stable/index.html). [here is an example of what that would look like!](https://geopandas.org/en/stable/gallery/index.html). if i can create the correct data subset i can then make tutorials that run when our docs build. We'd now be testing full workflows. we could then more easily address whether a few small wrappers (that i'm sure we're all creating to support these workflows) make sense as additions to stravalib! (ie creating specific use-cases that we identify as commonly shared)

### 6 months out?
Add [Doctest](https://docs.python.org/3/library/doctest.html): optional if we implement 2 above with a dataset we could then add testable examples to our object docstrings (id say this is a future item)

## Next Steps 
I'd like to focus on improved API documentation (docstrings), code format and tutorials. BUT i need to better understand how much the pydantic move will change the code as perhaps i wait on manual updates.  It would also be helpful to understand what the general timeline might look like for that move. it sounds like a lot of work and time!

Once the above is complete, i'm very interested in helping with some of the smaller future additions but i'm guessing those are probably 6+ months out! 

Just my thoughts! (with some questions mixed in!

## Docstrings and code format:
I could submit a pr that introduces pre-commit hooks, updates CI with format steps. and runs black on all modules. that would be doable and not a loss of time if the API ends up changing in structure via pydantic.

If the entire API changes, focusing on docstring or flake8 error updates (all manual work) might be a poor use of my time now. maybe i do that after the pydantic implementation of the api? 

Tutorials - i could easily start on those in the upcoming months. i think they can drive future feature additions as well. 

## Timing! 
Good to know about your time! maybe we figure out what our cadence is and maybe it becomes more like several times a week or once a week or whatever makes sense vs the more rapid development we've been doing in past weeks?

**Comment by jsamoocha** [comment]
Thanks for sharing your ideas!

Some of the topics you mention (e.g., API documentation, especially docstrings) will be impacted by the move to pydantic, especially since the data model code generator can also generate docstrings (from whatever is being published by Strava). We'd rather wait to work on that aspect until the pydantic change is finished and stable.

Adding tutorials sounds great! Would you happen to have a particular example or endpoint in mind regarding tutorial documentation? It may be a useful exercise to browse existing (closed) issues with "how to ..." questions to better understand the demand from the community. With "creating a small dataset," do you mean a mocked API response from Strava, e.g., a list of activities?

Finally, about the timing of the pydantic change, I'll outline more details in #248.

**Comment by lwasser** [comment]
Ok all sounds good. 
So i just 
1. submitted a PR for the metadata update - we can remove setup.cfg via that update and are then complying with the newer peps for packaging!
2. I plan to work on implementing the pre-commit hooks / linter / formatter changes next. I will need to coordinate with your work on the model for this change. Perhaps i just implement isort and hold off on black and flake8 but i add all of teh tooling for the pre commit hooks? i'll followup [in this pr](https://github.com/stravalib/stravalib/issues/293) on those specifics if you can check in there. 
3. the first tutorial i want to make/  improve is this one - https://stravalib.readthedocs.io/en/latest/get-started/authenticate-with-strava.html i think it's nice but setting up an app in your account should be included there. And i'd like to add some more detail about the token refresh. 

I created some little helper functions to implement all of that - that we could add there too. or perhaps you have an even better approach. but regardless i think being more clear about how to authenticate would be helpful. i know that people who are NOT developers would enjoy using this package! 

i actually created a pickle object for the secrets or a text file too to make it easier. 

ANd then YES i do want to create a small API return dataset to use in tutorials :) if we add sphinx gallery it will actually run the code beginning to end for each tutorial that we make! how about i create an issue for the first tutorial that i want to work on and we can chat more there?

**Comment by yihong0618** [comment]
Automatic request retrying (e.g., in case of temporary 5XX responses)
Automatic token refreshing
async requests? 🙈

---

1. sounds great
2. seems we can just use refresh token
3. that's a lot of work to do. and for now I think sync is fast enough?

**Comment by jsamoocha** [comment]
> seems we can just use refresh token

Do you mean we can use the refresh token as a permanent access token??

**Comment by jsamoocha** [comment]
> the first tutorial i want to make/ improve is this one - https://stravalib.readthedocs.io/en/latest/get-started/authenticate-with-strava.html i think it's nice but setting up an app in your account should be included there. And i'd like to add some more detail about the token refresh.

Sounds great, when looking at topics in the [strava API group](https://groups.google.com/g/strava-api), authentication is definitely something that many people are struggling with.

**Comment by yihong0618** [comment]
yes all my repos like [running_page](https://github.com/yihong0618/running_page) are using refresh_token and it runs three years/
